File: 프로그래머스 고득점 kit/Java/베스트앨범.java
import java.util.*;

class Solution {

    static class Music {
        String genre;
        int play;
        int idx;

        public Music(String genre, int play, int idx) {
            this.genre = genre;
            this.play = play;
            this.idx = idx;
        }
    }

    public int[] solution(String[] genres, int[] plays) {
        Map<String, Integer> genrePlay = new HashMap<>();
        int[] answer;

        for (int i = 0; i < genres.length; i++) {
            genrePlay.put(genres[i], genrePlay.getOrDefault(genres[i], 0) + plays[i]);
        }

        List<String> genreKeySet = new ArrayList<>(genrePlay.keySet());
        Collections.sort(genreKeySet, (value1, value2) -> (genrePlay.get(value2).compareTo(genrePlay.get(value1))));

        List<Music> result = new ArrayList<>();

        for (String genre : genreKeySet) {
            List<Music> tmp = new ArrayList<>();
            for (int i = 0; i < genres.length; i++) {
                if (genres[i].equals(genre)) {
                    tmp.add(new Music(genre, plays[i], i));
                }
            }

            Collections.sort(tmp, (o1, o2) -> o2.play - o1.play);
            result.add(tmp.get(0));
            if (tmp.size() != 1) {
                result.add(tmp.get(1));
            }
        }

        answer = new int[result.size()];
        for (int i = 0; i < result.size(); i++) {
            answer[i] = result.get(i).idx;
        }

        return answer;
    }
}

File: 프로그래머스 고득점 kit/C++/베스트앨범.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

bool compare_m2(pair<string, int> a, pair<string, int> b)
{
    return a.second > b.second;
}

bool compare_m(pair<int, int> a, pair<int, int> b)
{
    return a.first > b.first;
}

vector<int> solution(vector<string> genres, vector<int> plays)
{
    vector<int> result;
    unordered_map<string, vector<pair<int, int>>> m;
    unordered_map<string, int> m2;
    vector<pair<string, int>> genre_play;

    for (int i = 0; i < genres.size(); i++)
    {
        m2[genres[i]] += plays[i];
        m[genres[i]].push_back({plays[i], i});
    }

    for (auto &play : m)
    {
        sort(play.second.begin(), play.second.end(), compare_m);
    }

    genre_play.assign(m2.begin(), m2.end());
    sort(genre_play.begin(), genre_play.end(), compare_m2);

    for (int i = 0; i < genre_play.size(); i++)
    {
        string name = genre_play[i].first;
        for (int j = 0; (j < m[name].size()) && (j < 2); j++)
        {
            result.push_back(m[name][j].second);
        }
    }

    return result;
}